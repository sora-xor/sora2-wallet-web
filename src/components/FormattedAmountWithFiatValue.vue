<template>
  <div class="formatted-amount__container">
    <formatted-amount
      :class="valueClass"
      :value="value"
      :font-size-rate="fontSizeRate"
      :font-weight-rate="fontWeightRate"
      :asset-symbol="assetSymbol"
      :symbol-as-decimal="symbolAsDecimal"
      :value-can-be-hidden="valueCanBeHidden"
    >
      <slot />
    </formatted-amount>
    <formatted-amount
      v-if="hasFiatValue"
      is-fiat-value
      :value="fiatValue"
      :font-size-rate="fiatFormatAsValue ? fontSizeRate : fiatFontSizeRate"
      :font-weight-rate="fiatFormatAsValue ? fontWeightRate : fiatFontWeightRate"
      :with-left-shift="withLeftShift"
      :value-can-be-hidden="valueCanBeHidden"
    />
  </div>
</template>

<script lang="ts">
import { Vue, Component, Prop } from 'vue-property-decorator';

import { FontSizeRate, FontWeightRate } from '../consts';
import FormattedAmount from './FormattedAmount.vue';

@Component({
  components: {
    FormattedAmount,
  },
})
export default class FormattedAmountWithFiatValue extends Vue {
  /**
   * Amount value's custom class.
   */
  @Prop({ default: '', type: String }) readonly valueClass!: string;
  /**
   * Balance or Amount value.
   */
  @Prop({ default: '', type: String }) readonly value!: string;
  /**
   * Amount value's Font size rate between integer and decimal numbers' parts. Possible values: `"small"`, `"medium"`, `"normal"`.
   * By default it's set to `"normal"` and it means the same font sizes for both numbers' parts.
   */
  @Prop({ default: FontSizeRate.NORMAL, type: String }) readonly fontSizeRate!: string;
  /**
   * Amount value's Font weight rate between integer and decimal numbers' parts. Possible values: `"small"`, `"medium"`, `"normal"`.
   * By default it's set to `"normal"` and it means the same font weights for both numbers' parts.
   */
  @Prop({ default: FontWeightRate.NORMAL, type: String }) readonly fontWeightRate!: string;
  /**
   * Amount value's asset symbol.
   */
  @Prop({ default: '', type: String }) readonly assetSymbol!: string;
  /**
   * Font size of asset symbol is the same with decimal part size for Amount value.
   * Symbol value located inside formatted-amount__decimal container at the HTML structure.
   */
  @Prop({ default: false, type: Boolean }) readonly symbolAsDecimal!: boolean;
  /**
   * Sometimes presence of Fiat value related on some rules. With this setting we could show/hide Fiat value part.
   */
  @Prop({ default: true, type: Boolean }) readonly hasFiatValue!: boolean;
  /**
   * Fiat value.
   */
  @Prop({ default: '', type: String }) readonly fiatValue!: string;
  /**
   * We could set this flag to true if we have the same FontSizeRate and FontWeightRate with Amount value.
   */
  @Prop({ default: false, type: Boolean }) readonly fiatFormatAsValue!: boolean;
  /**
   * Define directly that this field displays value which can be hidden by hide balances button.
   */
  @Prop({ default: false, type: Boolean }) readonly valueCanBeHidden!: boolean;
  /**
   * Fiat value's Font size rate between integer and decimal numbers' parts. Possible values: `"small"`, `"medium"`, `"normal"`.
   * By default it's set to `"normal"` and it means the same font sizes for both numbers' parts.
   */
  @Prop({ default: FontSizeRate.NORMAL, type: String }) readonly fiatFontSizeRate!: string;
  /**
   * Fiat value's Font weight rate between integer and decimal numbers' parts. Possible values: `"small"`, `"medium"`, `"normal"`.
   * By default it's set to `"normal"` and it means the same font weights for both numbers' parts.
   */
  @Prop({ default: FontWeightRate.NORMAL, type: String }) readonly fiatFontWeightRate!: string;
  /**
   * Added special class to left shifting for Fiat value if needed (the shift is the same in all screens).
   */
  @Prop({ default: false, type: Boolean }) readonly withLeftShift!: boolean;
}
</script>

<style scoped lang="scss">
.formatted-amount__container {
  display: flex;
  justify-content: flex-end;
  align-items: baseline;
  flex-wrap: wrap;
  word-break: break-word;
  text-align: right;
}
</style>
