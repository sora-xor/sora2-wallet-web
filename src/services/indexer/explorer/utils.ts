import { formatStringNumber } from '../../../util';

import type { SubqueryAssetEntity } from '../subquery/types';
import type { SubsquidAssetEntity } from '../subsquid/types';
import type { FiatPriceObject, UpdatesStream, PoolApyObject } from '../types';

export function parseAssetFiatPrice(entity: SubsquidAssetEntity | SubqueryAssetEntity): FiatPriceObject {
  const acc = {};
  const id = entity.id;
  const priceFPNumber = formatStringNumber(entity.priceUSD);
  const isPriceFinity = priceFPNumber.isFinity();
  if (isPriceFinity) {
    acc[id] = priceFPNumber.toCodecString();
  }
  return acc;
}

export function parsePriceStreamUpdate(entity: UpdatesStream): Nullable<FiatPriceObject> {
  if (!entity?.data) return null;

  const data = JSON.parse(entity.data);

  return Object.entries(data).reduce((acc, [id, price]) => {
    const priceFPNumber = formatStringNumber(price as string);
    const isPriceFinity = priceFPNumber.isFinity();
    if (isPriceFinity) {
      acc[id] = priceFPNumber.toCodecString();
    }
    return acc;
  }, {});
}

export function parseApyStreamUpdate(entity: UpdatesStream): PoolApyObject {
  const data = entity?.data ? JSON.parse(entity.data) : {};

  return Object.entries(data).reduce((acc, [id, apy]) => {
    const strategicBonusApyFPNumber = formatStringNumber(apy as string);
    const isStrategicBonusApyFinity = strategicBonusApyFPNumber.isFinity();
    if (isStrategicBonusApyFinity) {
      acc[id] = strategicBonusApyFPNumber.toCodecString();
    }
    return acc;
  }, {});
}
