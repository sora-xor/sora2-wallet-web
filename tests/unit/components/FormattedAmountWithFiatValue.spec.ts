import { shallowMount } from '@vue/test-utils';

import { useDescribe, localVue } from '../../utils';
import { MOCK_FORMATTED_AMOUNT_WITH_FIAT_VALUE } from '../../utils/FormattedAmountWithFiatValueMock';

import FormattedAmountWithFiatValue from '@/components/FormattedAmountWithFiatValue.vue';

useDescribe('FormattedAmountWithFiatValue.vue', FormattedAmountWithFiatValue, () => {
  MOCK_FORMATTED_AMOUNT_WITH_FIAT_VALUE.map((item) =>
    it(`[${item.title}]: should be rendered correctly`, () => {
      const propsData = {
        valueClass: item.valueClass,
        value: item.value,
        fontSizeRate: item.fontSizeRate,
        fontWeightRate: item.fontWeightRate,
        assetSymbol: item.assetSymbol,
        symbolAsDecimal: item.symbolAsDecimal,
        hasFiatValue: item.hasFiatValue,
        fiatValue: item.fiatValue,
        fiatFormatAsValue: item.fiatFormatAsValue,
        fiatFontSizeRate: item.fiatFontSizeRate,
        fiatFontWeightRate: item.fiatFontWeightRate,
        withLeftShift: item.withLeftShift,
        valueCanBeHidden: item.valueCanBeHidden,
      };
      const wrapper = shallowMount(FormattedAmountWithFiatValue, {
        localVue,
        propsData,
      });
      expect(wrapper.element).toMatchSnapshot();
    })
  );
});
