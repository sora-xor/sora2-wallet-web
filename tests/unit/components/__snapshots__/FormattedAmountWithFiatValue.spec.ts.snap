// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormattedAmountWithFiatValue.vue [Fiat FontSizeRate Medium]: should be rendered correctly 1`] = `
<div
  class="formatted-amount__container"
>
  <formatted-amount-stub
    assetsymbol=""
    class=""
    fontsizerate="normal"
    fontweightrate="normal"
    value="1,234,567.890123"
  />
   
  <formatted-amount-stub
    assetsymbol=""
    fontsizerate="medium"
    fontweightrate="normal"
    isfiatvalue="true"
    value="1,234,567.890123"
  />
</div>
`;

exports[`FormattedAmountWithFiatValue.vue [Fiat FontWeightRate Medium]: should be rendered correctly 1`] = `
<div
  class="formatted-amount__container"
>
  <formatted-amount-stub
    assetsymbol=""
    class=""
    fontsizerate="normal"
    fontweightrate="normal"
    value="1,234,567.890123"
  />
   
  <formatted-amount-stub
    assetsymbol=""
    fontsizerate="normal"
    fontweightrate="medium"
    isfiatvalue="true"
    value="1,234,567.890123"
  />
</div>
`;

exports[`FormattedAmountWithFiatValue.vue [Fiat Format as Value Format Medium]: should be rendered correctly 1`] = `
<div
  class="formatted-amount__container"
>
  <formatted-amount-stub
    assetsymbol=""
    class=""
    fontsizerate="medium"
    fontweightrate="medium"
    value="1,234,567.890123"
  />
   
  <formatted-amount-stub
    assetsymbol=""
    fontsizerate="medium"
    fontweightrate="medium"
    isfiatvalue="true"
    value="1,234,567.890123"
  />
</div>
`;

exports[`FormattedAmountWithFiatValue.vue [FontSizeRate Small]: should be rendered correctly 1`] = `
<div
  class="formatted-amount__container"
>
  <formatted-amount-stub
    assetsymbol=""
    class=""
    fontsizerate="small"
    fontweightrate="normal"
    value="1,234,567.890123"
  />
   
  <formatted-amount-stub
    assetsymbol=""
    fontsizerate="normal"
    fontweightrate="normal"
    isfiatvalue="true"
    value="1,234,567.890123"
  />
</div>
`;

exports[`FormattedAmountWithFiatValue.vue [FontWeightRate Small]: should be rendered correctly 1`] = `
<div
  class="formatted-amount__container"
>
  <formatted-amount-stub
    assetsymbol=""
    class=""
    fontsizerate="normal"
    fontweightrate="small"
    value="1,234,567.890123"
  />
   
  <formatted-amount-stub
    assetsymbol=""
    fontsizerate="normal"
    fontweightrate="normal"
    isfiatvalue="true"
    value="1,234,567.890123"
  />
</div>
`;

exports[`FormattedAmountWithFiatValue.vue [Has not Fiat Value]: should be rendered correctly 1`] = `
<div
  class="formatted-amount__container"
>
  <formatted-amount-stub
    assetsymbol=""
    class=""
    fontsizerate="normal"
    fontweightrate="normal"
    value="1,234,567.890123"
  />
   
  <!---->
</div>
`;

exports[`FormattedAmountWithFiatValue.vue [Symbol In Decimal Container]: should be rendered correctly 1`] = `
<div
  class="formatted-amount__container"
>
  <formatted-amount-stub
    assetsymbol="XOR"
    class=""
    fontsizerate="normal"
    fontweightrate="normal"
    symbolasdecimal="true"
    value="1,234,567.890123"
  />
   
  <formatted-amount-stub
    assetsymbol=""
    fontsizerate="normal"
    fontweightrate="normal"
    isfiatvalue="true"
    value="1,234,567.890123"
  />
</div>
`;

exports[`FormattedAmountWithFiatValue.vue [Value and Fiat value only, can be hidden]: should be rendered correctly 1`] = `
<div
  class="formatted-amount__container"
>
  <formatted-amount-stub
    assetsymbol=""
    class=""
    fontsizerate="normal"
    fontweightrate="normal"
    value="1,234,567.890123"
    valuecanbehidden="true"
  />
   
  <formatted-amount-stub
    assetsymbol=""
    fontsizerate="normal"
    fontweightrate="normal"
    isfiatvalue="true"
    value="1,234,567.890123"
    valuecanbehidden="true"
  />
</div>
`;

exports[`FormattedAmountWithFiatValue.vue [Value and Fiat value only]: should be rendered correctly 1`] = `
<div
  class="formatted-amount__container"
>
  <formatted-amount-stub
    assetsymbol=""
    class=""
    fontsizerate="normal"
    fontweightrate="normal"
    value="1,234,567.890123"
  />
   
  <formatted-amount-stub
    assetsymbol=""
    fontsizerate="normal"
    fontweightrate="normal"
    isfiatvalue="true"
    value="1,234,567.890123"
  />
</div>
`;

exports[`FormattedAmountWithFiatValue.vue [Value only, can be hidden]: should be rendered correctly 1`] = `
<div
  class="formatted-amount__container"
>
  <formatted-amount-stub
    assetsymbol=""
    class=""
    fontsizerate="normal"
    fontweightrate="normal"
    value="1,234,567.890123"
    valuecanbehidden="true"
  />
   
  <formatted-amount-stub
    assetsymbol=""
    fontsizerate="normal"
    fontweightrate="normal"
    isfiatvalue="true"
    value=""
    valuecanbehidden="true"
  />
</div>
`;

exports[`FormattedAmountWithFiatValue.vue [Value only]: should be rendered correctly 1`] = `
<div
  class="formatted-amount__container"
>
  <formatted-amount-stub
    assetsymbol=""
    class=""
    fontsizerate="normal"
    fontweightrate="normal"
    value="1,234,567.890123"
  />
   
  <formatted-amount-stub
    assetsymbol=""
    fontsizerate="normal"
    fontweightrate="normal"
    isfiatvalue="true"
    value=""
  />
</div>
`;

exports[`FormattedAmountWithFiatValue.vue [With Empty Value]: should be rendered correctly 1`] = `
<div
  class="formatted-amount__container"
>
  <formatted-amount-stub
    assetsymbol=""
    class=""
    fontsizerate="normal"
    fontweightrate="normal"
    value=""
  />
   
  <formatted-amount-stub
    assetsymbol=""
    fontsizerate="normal"
    fontweightrate="normal"
    isfiatvalue="true"
    value=""
  />
</div>
`;

exports[`FormattedAmountWithFiatValue.vue [With Left Shift for Fiat]: should be rendered correctly 1`] = `
<div
  class="formatted-amount__container"
>
  <formatted-amount-stub
    assetsymbol=""
    class=""
    fontsizerate="normal"
    fontweightrate="normal"
    value="1,234,567.890123"
  />
   
  <formatted-amount-stub
    assetsymbol=""
    fontsizerate="normal"
    fontweightrate="normal"
    isfiatvalue="true"
    value="1,234,567.890123"
    withleftshift="true"
  />
</div>
`;

exports[`FormattedAmountWithFiatValue.vue [With asset symbol]: should be rendered correctly 1`] = `
<div
  class="formatted-amount__container"
>
  <formatted-amount-stub
    assetsymbol="XOR"
    class=""
    fontsizerate="normal"
    fontweightrate="normal"
    value="1,234,567.890123"
  />
   
  <formatted-amount-stub
    assetsymbol=""
    fontsizerate="normal"
    fontweightrate="normal"
    isfiatvalue="true"
    value="1,234,567.890123"
  />
</div>
`;

exports[`FormattedAmountWithFiatValue.vue [With value class]: should be rendered correctly 1`] = `
<div
  class="formatted-amount__container"
>
  <formatted-amount-stub
    assetsymbol=""
    class="formatted-amount--test"
    fontsizerate="normal"
    fontweightrate="normal"
    value="1,234,567.890123"
  />
   
  <formatted-amount-stub
    assetsymbol=""
    fontsizerate="normal"
    fontweightrate="normal"
    isfiatvalue="true"
    value="1,234,567.890123"
  />
</div>
`;
